File: apps/windows/TeslaSync.App/feature-views/MiniGridPreview/MiniGridPreview.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Automation.Peers;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using TeslaSync.App.Components.DataDisplay;
using TeslaSync.App.Components.Feedback;
using TeslaSync.App.Components.UI;
using TeslaSync.App.Core.Notifications;
using Windows.Foundation;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The native WinUI 3 <c>MiniGridPreview</c> feature surface — a parity port of
/// web/src/features/dashboard/components/MiniGridPreview.tsx. It renders a scaled-down thumbnail of a saved
/// dashboard's <c>lg</c> grid: a tokenized rounded container (web <c>bg-white/[0.02]</c> +
/// <c>border-white/[0.06]</c> + <c>rounded-lg</c>, clipped like web <c>overflow-hidden</c>) whose
/// width-to-height ratio is locked to <c>cols / safeMaxY</c> (web <c>aspectRatio</c>), holding one absolutely
/// positioned cell per placed widget (web <c>bg-white/[0.06]</c> + <c>border-white/[0.08]</c> +
/// <c>rounded-sm</c>) centred on the widget's catalog icon at 12px in the muted token (web <c>Icon</c> at
/// <c>h-3 w-3 text-[var(--text-muted)]</c>). The surface is purely presentational and fetches nothing, so —
/// like the web source — it has no loading / error / stale / offline branch; the only content states are
/// populated (cells laid out from the projection) and empty (a friendly centred stand-in, never a blank
/// box). All UI-free decisions (geometry, the icon join, copy resolution) live in
/// <see cref="MiniGridPreviewProjection"/> / <see cref="MiniGridWidgetIcons"/> so they are verified without a
/// UI host. The preview carries a single localized Narrator name; the decorative cells and glyphs are hidden
/// from Narrator, and the empty state announces through a polite live region.
/// </summary>
public sealed partial class MiniGridPreview : ContentControl
{
    private const double ContainerRadius = 8;     // web rounded-lg
    private const double TileRadius = 2;          // web rounded-sm
    private const double TilePadding = 2;         // web padding: '2px'
    private const double TileBorderThickness = 1; // web border
    private const double IconSize = 12;           // web h-3 w-3
    private const double EmptyIconSize = 24;
    private const double EmptySpacing = 8;
    private const double EmptyPadding = 12;
    private const double AspectEpsilon = 0.5;
    private const string GlassBrushKey = "TsColorSurfaceGlassBrush";

    private readonly ILocalizer _localizer;
    private readonly MiniGridPreviewDiagnostics _diagnostics;
    private readonly Border _container;
    private readonly Canvas _canvas = new();
    private readonly List<(FrameworkElement Element, MiniGridTile Tile)> _tiles = new();

    private MiniGridPreviewModel _model;
    private MiniGridPreviewDisplay _display;
    private FrameworkElement? _liveRegion;
    private bool _opened;

    /// <summary>Creates the surface over its i18n facade, an initial model, and (optional) diagnostics.</summary>
    /// <param name="localizer">The i18n facade the accessible name / empty caption resolve through.</param>
    /// <param name="model">The initial render model; defaults to <see cref="MiniGridPreviewModel.Empty"/>.</param>
    /// <param name="diagnostics">The PII-safe diagnostics sink for the <c>view.opened</c> event.</param>
    public MiniGridPreview(
        ILocalizer localizer,
        MiniGridPreviewModel? model = null,
        MiniGridPreviewDiagnostics? diagnostics = null)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        _localizer = localizer;
        _model = model ?? MiniGridPreviewModel.Empty;
        _diagnostics = diagnostics ?? new MiniGridPreviewDiagnostics();
        _display = MiniGridPreviewProjection.Project(_model, _localizer);

        IsTabStop = false;
        HorizontalAlignment = HorizontalAlignment.Stretch;
        VerticalAlignment = VerticalAlignment.Top;
        HorizontalContentAlignment = HorizontalAlignment.Stretch;
        VerticalContentAlignment = VerticalAlignment.Top;

        _container = new Border
        {
            Background = DisplayTokens.Brush(GlassBrushKey),
            BorderBrush = DisplayTokens.Border,
            BorderThickness = new Thickness(1),
            CornerRadius = new CornerRadius(ContainerRadius),
            HorizontalAlignment = HorizontalAlignment.Stretch,
            VerticalAlignment = VerticalAlignment.Top,
        };
        _container.SizeChanged += OnContainerSizeChanged;

        Content = _container;
        Loaded += OnLoaded;
        Render();
    }

    /// <summary>The canonical diagnostics slug this surface registers under (<c>MiniGridPreview</c>).</summary>
    public static string Slug => MiniGridPreviewRegistration.Slug;

    /// <summary>The render model; reassigning re-projects and re-renders the surface.</summary>
    public MiniGridPreviewModel Model
    {
        get => _model;
        set
        {
            ArgumentNullException.ThrowIfNull(value);
            _model = value;
            _display = MiniGridPreviewProjection.Project(_model, _localizer);
            Render();
        }
    }

    private void OnLoaded(object sender, RoutedEventArgs e)
    {
        if (_liveRegion is not null)
        {
            LiveRegion.Announce(_liveRegion);
        }

        if (_opened)
        {
            return;
        }

        _opened = true;
        _diagnostics.RecordViewOpened();
    }

    private void Render()
    {
        _tiles.Clear();
        _canvas.Children.Clear();
        _liveRegion = null;

        AutomationProperties.SetName(this, _display.AutomationName);
        AutomationProperties.SetName(_container, _display.AutomationName);

        if (_display.IsEmpty)
        {
            // web renders an empty bordered box; we keep the box but never leave it blank.
            _container.Child = BuildEmpty();
        }
        else
        {
            foreach (var tile in _display.Tiles)
            {
                var element = BuildTile(tile);
                _tiles.Add((element, tile));
                _canvas.Children.Add(element);
            }

            _container.Child = _canvas;
        }

        if (_container.ActualWidth > 0)
        {
            ApplyLayout(_container.ActualWidth);
        }
    }

    // ── One cell: a tokenized rounded rectangle centred on its (decorative) widget glyph ─────────────────
    private static Border BuildTile(MiniGridTile tile)
    {
        var border = new Border
        {
            Background = DisplayTokens.Brush(GlassBrushKey),
            BorderBrush = DisplayTokens.Border,
            BorderThickness = new Thickness(TileBorderThickness),
            CornerRadius = new CornerRadius(TileRadius),
            Padding = new Thickness(TilePadding),
        };

        if (tile.IconGlyph is { } glyph)
        {
            var icon = new FontIcon
            {
                Glyph = glyph,
                FontSize = IconSize,
                Foreground = DisplayTokens.TextMuted,
                HorizontalAlignment = HorizontalAlignment.Center,
                VerticalAlignment = VerticalAlignment.Center,
            };

            // Decorative — the preview's single Narrator name carries the meaning (web aria-hidden).
            AutomationProperties.SetAccessibilityView(icon, AccessibilityView.Raw);
            border.Child = icon;
        }

        AutomationProperties.SetAccessibilityView(border, AccessibilityView.Raw);
        return border;
    }

    // ── Empty state: a centred glyph + localized caption, announced politely (never a blank box) ─────────
    private StackPanel BuildEmpty()
    {
        var stack = new StackPanel
        {
            Orientation = Orientation.Vertical,
            HorizontalAlignment = HorizontalAlignment.Center,
            VerticalAlignment = VerticalAlignment.Center,
            Spacing = EmptySpacing,
            Padding = new Thickness(EmptyPadding),
        };

        var icon = new FontIcon
        {
            Glyph = MiniGridPreviewRegistration.EmptyGlyph,
            FontSize = EmptyIconSize,
            Foreground = DisplayTokens.TextMuted,
            HorizontalAlignment = HorizontalAlignment.Center,
        };
        AutomationProperties.SetAccessibilityView(icon, AccessibilityView.Raw);
        stack.Children.Add(icon);

        stack.Children.Add(new TextBlock
        {
            Text = _display.EmptyMessage,
            FontSize = TypographyTokens.Size("TsTypeCaptionFontSize", 12),
            Foreground = DisplayTokens.TextMuted,
            HorizontalAlignment = HorizontalAlignment.Center,
            TextAlignment = TextAlignment.Center,
            TextWrapping = TextWrapping.Wrap,
        });

        AutomationProperties.SetName(stack, _display.EmptyMessage);
        LiveRegion.Configure(stack);
        _liveRegion = stack;
        return stack;
    }

    private void OnContainerSizeChanged(object sender, SizeChangedEventArgs e) => ApplyLayout(e.NewSize.Width);

    // Lock the container to the web aspect ratio (width : height = cols : safeMaxY) and position every cell
    // from its 0..1 fractions against the rendered size, clipping to the bounds (web overflow-hidden).
    private void ApplyLayout(double width)
    {
        if (width <= 0 || double.IsNaN(width) || double.IsInfinity(width))
        {
            return;
        }

        double height = width * _display.RowSpan / _display.Columns;
        if (double.IsNaN(_container.Height) || Math.Abs(_container.Height - height) > AspectEpsilon)
        {
            _container.Height = height; // re-fires SizeChanged; the epsilon guard prevents a feedback loop
        }

        _canvas.Clip = new RectangleGeometry { Rect = new Rect(0, 0, width, height) };

        foreach (var (element, tile) in _tiles)
        {
            Canvas.SetLeft(element, tile.LeftFraction * width);
            Canvas.SetTop(element, tile.TopFraction * height);
            element.Width = Math.Max(0, tile.WidthFraction * width);
            element.Height = Math.Max(0, tile.HeightFraction * height);
        }
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/MiniGridPreview/MiniGridPreview.kt
// The native Jetpack Compose + Material 3 `MiniGridPreview` feature view — a parity port of
// web/src/features/dashboard/components/MiniGridPreview.tsx. The web component is a tiny presentational thumbnail:
// given a `SavedDashboard`, it draws that dashboard's `lg` react-grid-layout as a four-column frame of small boxes,
// each box positioned by percentage and carrying its widget-registry icon centered inside. It fetches nothing and
// renders no text — purely a "shape of this layout" glance used in switchers/preset galleries.
//
// This port keeps that composition end to end and performs NO HTTP. All of the geometry the web derives lives in the
// pure [io.teslasync.android.featureviews.minigridpreview.MiniGridPreviewProjection] (the row-count guard, the
// `cols / safeMaxY` aspect ratio, and each box's fractional rectangle), unit-tested off-device, so this file is a
// thin render layer. Resolving a widget id to an icon is the widget-registry surface's concern (a separate prompt,
// exactly as the sibling `DashboardGrid` treats it), so the host supplies an [iconForWidget] resolver; a `null`
// result paints an empty box, mirroring the web `{Icon && <Icon/>}` guard.
//
// The web component receives a ready `dashboard` prop, so a web-parity overload taking a [MiniGridDashboard] is
// provided. Because the saved-dashboard feed is host-owned (P1/S8), the canonical entry takes a [UiState] and
// renders every lifecycle state that layer can carry — loading (skeleton), hard error (retry), the host-feed empty
// ("no dashboard" — a friendly state), content (the thumbnail), and stale/offline (the cached thumbnail plus an
// Offline freshness chip that auto-refreshes). A dashboard whose `lg` layout is empty is NOT the empty STATE: the
// web draws the bordered empty frame for it, so this port does too (content, never a blank box).
//
// Colors/typography map to design tokens (never raw hex / ad-hoc text styles): the frame and cell fills are subtle
// `onSurface` alpha overlays (the theme-adaptive equivalent of the web `white/[0.02|0.06]`), borders use the
// `outlineVariant`/`onSurface` outline tokens, and the cell icons inherit the muted `onSurfaceVariant` foreground
// (the web `--text-muted`). The thumbnail is a decorative composite, so the whole frame is collapsed to a single
// localized accessibility node ([clearAndSetSemantics]) and the inner cells/icons are not separately announced.
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory
// (com/teslasync/feature-views/MiniGridPreview — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package, so the package intentionally diverges from the path. `MatchingDeclarationName` is suppressed for the
// co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.minigridpreview

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import io.teslasync.android.R
import io.teslasync.android.components.datadisplay.DataFreshness
import io.teslasync.android.components.datadisplay.FreshnessAge
import io.teslasync.android.components.feedback.EmptyState
import io.teslasync.android.components.feedback.ErrorDisplay
import io.teslasync.android.components.feedback.Skeleton
import io.teslasync.android.components.motion.FadeIn
import io.teslasync.android.components.ui.Icon
import io.teslasync.android.components.ui.IconSize
import io.teslasync.android.components.ui.TeslaGlyphs
import io.teslasync.android.data.ErrorKind
import io.teslasync.android.data.LocalDataContainer
import io.teslasync.android.data.UiPhase
import io.teslasync.android.data.UiState
import io.teslasync.android.ui.theme.TeslaSyncTheme
import io.teslasync.android.ui.theme.generated.Radius
import io.teslasync.android.ui.theme.generated.Spacing
import io.teslasync.shared.core.diagnostics.Logger

/** Thumbnail entrance fade — a gentle reveal that honours reduced motion (the app's motion language). */
private const val PREVIEW_FADE_MS: Int = 200

/** Loading-skeleton thumbnail height in dp — sized so the loading state reads as a card, never a blank box. */
private val LOADING_THUMBNAIL_HEIGHT: Dp = 72.dp

/** Outline width of the frame and each cell, in dp (the web 1px borders). */
private val HAIRLINE_WIDTH: Dp = 1.dp

/** Corner radius of a cell box, in dp — the web `rounded-sm` (2px). */
private val CELL_CORNER_RADIUS: Dp = 2.dp

/** Inner inset between a cell's border and its icon, in dp — the web cell `padding: '2px'`. */
private val CELL_INSET: Dp = 2.dp

/** Frame fill opacity over the theme foreground — the theme-adaptive equivalent of the web `bg-white/[0.02]`. */
private const val FRAME_FILL_ALPHA: Float = 0.02f

/** Cell fill opacity over the theme foreground — the web `bg-white/[0.06]`. */
private const val CELL_FILL_ALPHA: Float = 0.06f

/** Cell border opacity over the theme foreground — the web `border-white/[0.08]`. */
private const val CELL_BORDER_ALPHA: Float = 0.08f

/** Em dash shown when a freshness age is unknown — the shared freshness fallback. */
private const val EM_DASH: String = "\u2014"

/**
 * Stateful entry point for `MiniGridPreview`. Records the one-shot PII-safe `view.opened` diagnostic (P1/S11) and
 * renders every lifecycle [state] the saved-dashboard feed can carry. The host owns the feed (P1/S8) and supplies
 * [onRetry] (its `refetch`); this view never performs HTTP.
 *
 * @param state the cache-then-network projection of the [MiniGridDashboard] to preview.
 * @param onRetry re-runs the host's load — wired to the hard-error retry and the stale auto-refresh.
 * @param iconForWidget resolves a widget id to its registry icon (the web `getWidgetDef(widgetId)?.icon`); `null`
 *   paints an empty cell box. Defaults to no icons so the frame degrades gracefully without the registry.
 * @param logger the sanctioned redacting logger; defaults to the app's `LocalDataContainer`.
 */
@Composable
fun MiniGridPreview(
    state: UiState<MiniGridDashboard>,
    onRetry: () -> Unit,
    modifier: Modifier = Modifier,
    iconForWidget: (String) -> ImageVector? = { null },
    logger: Logger = LocalDataContainer.current.logger,
) {
    LaunchedEffect(Unit) { recordMiniGridPreviewOpened(logger) }
    MiniGridPreviewContent(
        state = state,
        onRetry = onRetry,
        modifier = modifier,
        iconForWidget = iconForWidget,
    )
}

/**
 * Web-parity overload mirroring the web component's ready `dashboard` prop, for hosts that already hold the saved
 * layout. Always renders the thumbnail (an empty `lg` layout draws the bordered empty frame, matching the web).
 * Records `view.opened` like the stateful entry; there is no fetch behind it, so it offers no retry affordance.
 */
@Composable
fun MiniGridPreview(
    dashboard: MiniGridDashboard,
    modifier: Modifier = Modifier,
    iconForWidget: (String) -> ImageVector? = { null },
    logger: Logger = LocalDataContainer.current.logger,
) {
    val state = remember(dashboard) { MiniGridPreviewProjection.contentState(dashboard) }
    MiniGridPreview(
        state = state,
        onRetry = {},
        modifier = modifier,
        iconForWidget = iconForWidget,
        logger = logger,
    )
}

/**
 * Stateless renderer for every surface state — the unit/UI-test entry point. Loading shows a skeleton thumbnail; a
 * hard error shows a retry surface; the host-feed empty shows a friendly empty state; otherwise the thumbnail is
 * drawn (an empty layout still draws the bordered frame). Stale (non-error) data auto-refreshes, mirroring the web
 * freshness contract.
 */
@Composable
fun MiniGridPreviewContent(
    state: UiState<MiniGridDashboard>,
    onRetry: () -> Unit,
    modifier: Modifier = Modifier,
    iconForWidget: (String) -> ImageVector? = { null },
    strings: MiniGridPreviewStrings = rememberMiniGridPreviewStrings(),
) {
    LaunchedEffect(state.stale, state.refreshing, state.hasError) {
        if (state.stale && !state.refreshing && !state.hasError) onRetry()
    }
    when (miniGridPreviewSurface(state)) {
        MiniGridPreviewSurface.Loading -> MiniGridPreviewLoading(modifier = modifier.fillMaxWidth())
        MiniGridPreviewSurface.Error ->
            ErrorDisplay(
                message = strings.errorMessage,
                modifier = modifier.fillMaxWidth(),
                onRetry = onRetry,
                retryLabel = strings.retryLabel,
            )
        MiniGridPreviewSurface.Empty ->
            EmptyState(message = strings.emptyMessage, modifier = modifier.fillMaxWidth())
        MiniGridPreviewSurface.Content ->
            MiniGridPreviewScaffold(
                state = state,
                strings = strings,
                iconForWidget = iconForWidget,
                modifier = modifier,
            )
    }
}

/**
 * The content scaffold for the thumbnail: projects the layout, shows the freshness chip when cached data is
 * refreshing/stale/offline, and draws the frame. The frame is wrapped in a reduced-motion-aware [FadeIn].
 */
@Composable
private fun MiniGridPreviewScaffold(
    state: UiState<MiniGridDashboard>,
    strings: MiniGridPreviewStrings,
    iconForWidget: (String) -> ImageVector?,
    modifier: Modifier = Modifier,
) {
    val dashboard = state.data ?: MiniGridDashboard()
    val projection = remember(dashboard) { MiniGridPreviewProjection.project(dashboard) }
    val showFreshness = state.refreshing || state.stale || state.hasError

    FadeIn(durationMs = PREVIEW_FADE_MS) {
        Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(Spacing.xs)) {
            if (showFreshness) {
                Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
                    MiniGridPreviewFreshnessChip(state = state, strings = strings)
                }
            }
            MiniGridPreviewFrame(
                projection = projection,
                iconForWidget = iconForWidget,
                accessibilityLabel = strings.previewLabel,
            )
        }
    }
}

/**
 * The bordered preview frame: a [BoxWithConstraints] (the `useContainerWidth`-free measurement the web gets from CSS
 * percentages) laid out at the projection's aspect ratio, with each placed cell positioned by its fractional
 * rectangle. Collapsed to a single localized accessibility node so the decorative thumbnail is announced once.
 */
@Composable
private fun MiniGridPreviewFrame(
    projection: MiniGridPreviewProjectionResult,
    iconForWidget: (String) -> ImageVector?,
    accessibilityLabel: String,
    modifier: Modifier = Modifier,
) {
    val frameShape = RoundedCornerShape(Radius.sm)
    val mutedTint = MaterialTheme.colorScheme.onSurfaceVariant
    BoxWithConstraints(
        modifier =
            modifier
                .fillMaxWidth()
                .aspectRatio(projection.aspectRatio)
                .clip(frameShape)
                .background(MaterialTheme.colorScheme.onSurface.copy(alpha = FRAME_FILL_ALPHA))
                .border(HAIRLINE_WIDTH, MaterialTheme.colorScheme.outlineVariant, frameShape)
                .clearAndSetSemantics { contentDescription = accessibilityLabel },
    ) {
        val frameWidth = maxWidth
        val frameHeight = maxHeight
        projection.cells.forEach { cell ->
            MiniGridCellBox(
                cell = cell,
                frameWidth = frameWidth,
                frameHeight = frameHeight,
                icon = cell.widgetId?.let(iconForWidget),
                mutedTint = mutedTint,
            )
        }
    }
}

/**
 * One placed widget box — positioned and sized by multiplying the cell's fractional rectangle by the measured
 * [frameWidth]/[frameHeight] (the web absolute `left/top/width/height` percentages), with the widget [icon] centered
 * inside (omitted when the registry resolved nothing, the web `find`/`def` miss).
 */
@Composable
private fun MiniGridCellBox(
    cell: MiniGridCell,
    frameWidth: Dp,
    frameHeight: Dp,
    icon: ImageVector?,
    mutedTint: Color,
) {
    val cellShape = RoundedCornerShape(CELL_CORNER_RADIUS)
    val foreground = MaterialTheme.colorScheme.onSurface
    Box(
        modifier =
            Modifier
                .offset(x = frameWidth * cell.leftFraction, y = frameHeight * cell.topFraction)
                .size(width = frameWidth * cell.widthFraction, height = frameHeight * cell.heightFraction)
                .clip(cellShape)
                .background(foreground.copy(alpha = CELL_FILL_ALPHA))
                .border(HAIRLINE_WIDTH, foreground.copy(alpha = CELL_BORDER_ALPHA), cellShape)
                .padding(CELL_INSET),
        contentAlignment = Alignment.Center,
    ) {
        if (icon != null) {
            Icon(icon, contentDescription = null, size = IconSize.Xs, tint = mutedTint)
        }
    }
}

/** First-load skeleton — a single shimmering thumbnail-shaped block so the surface is never blank while loading. */
@Composable
private fun MiniGridPreviewLoading(modifier: Modifier = Modifier) {
    Skeleton(modifier = modifier.fillMaxWidth(), height = LOADING_THUMBNAIL_HEIGHT, rounded = false)
}

/**
 * The freshness chip shown above the thumbnail when cached data is refreshing / stale / offline — the honest "last
 * known + retry" affordance. Offline (a failed refresh over cached data) reads the localized "Offline" label; a
 * stale-but-reachable value reads its relative age. Mirrors the sibling surfaces' contract.
 */
@Composable
private fun MiniGridPreviewFreshnessChip(
    state: UiState<*>,
    strings: MiniGridPreviewStrings,
) {
    DataFreshness(
        updatedAtMillis = state.fetchedAt?.takeIf { it > 0 },
        isFetching = state.refreshing,
        isStale = state.stale,
        isError = state.hasError,
        compact = true,
        fetchingLabel = strings.loadingLabel,
        errorLabel = strings.offlineLabel,
        formatAge = rememberMiniGridPreviewFreshnessFormatter(),
    )
}

/**
 * Builds the localized [MiniGridPreviewStrings] from the i18n catalog (P1/S10): the decorative thumbnail's
 * accessible name (`app.preview`), the host-feed empty message (`common.noData`), and the lifecycle chrome
 * (error/retry/loading/offline). Remembered against the resolved strings so a locale change re-reads them.
 */
@Composable
fun rememberMiniGridPreviewStrings(): MiniGridPreviewStrings {
    val previewLabel = stringResource(R.string.translation_app_preview)
    val emptyMessage = stringResource(R.string.translation_common_noData)
    val errorMessage = stringResource(R.string.translation_error_serverError_message)
    val retryLabel = stringResource(R.string.translation_common_retry)
    val loadingLabel = stringResource(R.string.translation_common_loading)
    val offlineLabel = stringResource(R.string.translation_common_offline)
    return remember(previewLabel, emptyMessage, errorMessage, retryLabel, loadingLabel, offlineLabel) {
        MiniGridPreviewStrings(
            previewLabel = previewLabel,
            emptyMessage = emptyMessage,
            errorMessage = errorMessage,
            retryLabel = retryLabel,
            loadingLabel = loadingLabel,
            offlineLabel = offlineLabel,
        )
    }
}

/**
 * Localized relative-age formatter for the freshness chip (`translation_freshness_*`) — the same render-only concern
 * the sibling surfaces resolve, kept out of the pure projection.
 */
@Composable
private fun rememberMiniGridPreviewFreshnessFormatter(): (FreshnessAge) -> String {
    val justNow = stringResource(R.string.translation_freshness_justNow)
    val seconds = stringResource(R.string.translation_freshness_seconds)
    val minutes = stringResource(R.string.translation_freshness_minutes)
    val hours = stringResource(R.string.translation_freshness_hours)
    val days = stringResource(R.string.translation_freshness_days)
    val weeks = stringResource(R.string.translation_freshness_weeks)
    return remember(justNow, seconds, minutes, hours, days, weeks) {
        { age ->
            when (age) {
                FreshnessAge.Unknown -> EM_DASH
                FreshnessAge.JustNow -> justNow
                is FreshnessAge.Seconds -> seconds.format(age.value)
                is FreshnessAge.Minutes -> minutes.format(age.value)
                is FreshnessAge.Hours -> hours.format(age.value)
                is FreshnessAge.Days -> days.format(age.value)
                is FreshnessAge.Weeks -> weeks.format(age.value)
            }
        }
    }
}

/**
 * The already-localized microcopy the composable reads (P1/S10): the thumbnail's accessible name and the lifecycle
 * chrome. Kept out of the pure projection so it stays UI-free.
 */
data class MiniGridPreviewStrings(
    val previewLabel: String,
    val emptyMessage: String,
    val errorMessage: String,
    val retryLabel: String,
    val loadingLabel: String,
    val offlineLabel: String,
)

// ── Previews (tooling-only; @Preview entry points exercise each render branch) ──────────────────────

private val PREVIEW_STRINGS =
    MiniGridPreviewStrings(
        previewLabel = "Preview",
        emptyMessage = "No data available",
        errorMessage = "Something went wrong on our end. Please try again.",
        retryLabel = "Retry",
        loadingLabel = "Loading...",
        offlineLabel = "Offline",
    )

private val PREVIEW_DASHBOARD =
    MiniGridDashboard(
        widgets =
            listOf(
                MiniGridWidget(id = "w-1", widgetId = "vehicle-hero"),
                MiniGridWidget(id = "w-2", widgetId = "battery-gauge"),
                MiniGridWidget(id = "w-3", widgetId = "range-bar"),
                MiniGridWidget(id = "w-4", widgetId = "fleet-stats"),
            ),
        lgLayout =
            listOf(
                MiniGridLayoutItem("w-1", x = 0, y = 0, w = 2, h = 2),
                MiniGridLayoutItem("w-2", x = 2, y = 0, w = 1, h = 2),
                MiniGridLayoutItem("w-3", x = 3, y = 0, w = 1, h = 1),
                MiniGridLayoutItem("w-4", x = 0, y = 2, w = 2, h = 1),
            ),
    )

private val PREVIEW_ICONS: Map<String, ImageVector> =
    mapOf(
        "vehicle-hero" to TeslaGlyphs.Eye,
        "battery-gauge" to TeslaGlyphs.Check,
        "range-bar" to TeslaGlyphs.Info,
        "fleet-stats" to TeslaGlyphs.Pin,
    )

private fun previewIcon(widgetId: String): ImageVector? = PREVIEW_ICONS[widgetId]

@Preview(name = "Loading", showBackground = true, widthDp = 240)
@Composable
private fun MiniGridPreviewLoadingPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        MiniGridPreviewContent(state = UiState(UiPhase.Loading), onRetry = {}, strings = PREVIEW_STRINGS)
    }
}

@Preview(name = "Empty (no dashboard)", showBackground = true, widthDp = 240)
@Composable
private fun MiniGridPreviewEmptyPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        MiniGridPreviewContent(state = UiState(UiPhase.Empty), onRetry = {}, strings = PREVIEW_STRINGS)
    }
}

@Preview(name = "Error", showBackground = true, widthDp = 240)
@Composable
private fun MiniGridPreviewErrorPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        MiniGridPreviewContent(
            state = UiState(UiPhase.Error, errorKind = ErrorKind.Network),
            onRetry = {},
            strings = PREVIEW_STRINGS,
        )
    }
}

@Preview(name = "Content (thumbnail)", showBackground = true, widthDp = 240)
@Composable
private fun MiniGridPreviewContentPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        MiniGridPreviewContent(
            state = UiState(UiPhase.Content, data = PREVIEW_DASHBOARD),
            onRetry = {},
            iconForWidget = ::previewIcon,
            strings = PREVIEW_STRINGS,
        )
    }
}

@Preview(name = "Content (empty layout frame)", showBackground = true, widthDp = 240)
@Composable
private fun MiniGridPreviewEmptyLayoutPreview() {
    TeslaSyncTheme(dynamicColor = false) {
        MiniGridPreviewContent(
            state = UiState(UiPhase.Content, data = MiniGridDashboard()),
            onRetry = {},
            strings = PREVIEW_STRINGS,
        )
    }
}

@Preview(name = "Offline (cached)", showBackground = true, widthDp = 240)
@Composable
private fun MiniGridPreviewOfflinePreview() {
    TeslaSyncTheme(dynamicColor = false) {
        MiniGridPreviewContent(
            state =
                UiState(
                    phase = UiPhase.Content,
                    data = PREVIEW_DASHBOARD,
                    stale = true,
                    fetchedAt = 1_700_000_000_000L,
                    errorKind = ErrorKind.Network,
                ),
            onRetry = {},
            iconForWidget = ::previewIcon,
            strings = PREVIEW_STRINGS,
        )
    }
}
